File: Java/UserAndGroupCurdSample/src/main/java/com/thoughtspot/UserAndGroupCurdSample/GroupMethods.java
package com.thoughtspot.UserAndGroupCurdSample;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import localhost.RESTAPISDKClient;
import localhost.controllers.GroupController;
import localhost.exceptions.ApiException;
import localhost.models.*;

/**
 * Group CRUD operations.
 */
public class GroupMethods {
    /**
     * Create a new group by name.
     *
     * @param client client.
     * @param name name.
     * @return group response.
     * @throws IOException .
     * @throws ApiException .
     */
    public GroupResponse createNewGroup(RESTAPISDKClient client,
                                        String name) throws IOException, ApiException {
        GroupController groupController = client.getGroupController();

        int status = 0;
        GroupResponse result = null;

        try {
            result = groupController.getGroup(name, "");
            status = 1;
        } catch (Exception e) {
        } finally {
            if (status == 0) {
                TspublicRestV2GroupCreateRequest tspublicRestV2GroupCreateRequest =
                        new TspublicRestV2GroupCreateRequest();

                tspublicRestV2GroupCreateRequest.setName(name);
                tspublicRestV2GroupCreateRequest.setDisplayName(name);
                tspublicRestV2GroupCreateRequest.setVisibility(CreateGroupVisibilityEnum.DEFAULT);
                List<CreateGroupPrivilegesEnum> privilegeEnum = new ArrayList<>();
                privilegeEnum.add(CreateGroupPrivilegesEnum.DEVELOPER);
                privilegeEnum.add(CreateGroupPrivilegesEnum.DATAMANAGEMENT);
                privilegeEnum.add(CreateGroupPrivilegesEnum.EXPERIMENTALFEATUREPRIVILEGE);
                tspublicRestV2GroupCreateRequest.setPrivileges(privilegeEnum);
                tspublicRestV2GroupCreateRequest.setType(CreateGroupTypeEnum.LOCAL_GROUP);

                result = groupController.createGroup(tspublicRestV2GroupCreateRequest);
            }
        }
        return result;
    }

    /**
     * Add user to a group.
     *
     * @param client client.
     * @param userNames userNames.
     * @param groupname groupname.
     * @return returns true if added else returns false.
     * @throws IOException .
     * @throws ApiException .
     */
    public Boolean addUserToGroup(RESTAPISDKClient client,
                                  List<TspublicRestV2UserCreateRequest> userNames,
                                  String groupname) throws IOException, ApiException {

        GroupController groupController = client.getGroupController();

        TspublicRestV2GroupAdduserRequest tspublicRestV2GroupAdduserRequest =
                new TspublicRestV2GroupAdduserRequest();
        tspublicRestV2GroupAdduserRequest.setName(groupname);
        List<UserNameAndIDInput> userList = new ArrayList<>();

        for (TspublicRestV2UserCreateRequest tspublicRestV2UserCreateRequest : userNames) {
            UserNameAndIDInput userNameAndIDInput = new UserNameAndIDInput();
            userNameAndIDInput.setName(tspublicRestV2UserCreateRequest.getName());
            userList.add(userNameAndIDInput);
        }
        tspublicRestV2GroupAdduserRequest.setUsers(userList);

        Boolean result = groupController.addUsersToGroup(tspublicRestV2GroupAdduserRequest);

        return result;
    }

    /**
     *
     * @param client RESTAPISDKClient
     * @param group group name and id
     * @return returns true when group deleted successfully.
     * @throws IOException .
     * @throws ApiException .
     */
    public Boolean deleteGroup(
            RESTAPISDKClient client, UserNameAndID group)
            throws IOException, ApiException {

        GroupController groupController = client.getGroupController();
        return groupController.deleteGroup(group.getName(), null);
    }
}

File: Python/UserAndGroupCrudSample/UserAndGroupCurd/GroupMethods.py
#! /usr/bin/env python3
# pylint: disable= no-name-in-module
from restapisdk.models.create_group_privileges_enum import CreateGroupPrivilegesEnum
from restapisdk.models.create_group_type_enum import CreateGroupTypeEnum
from restapisdk.models.create_group_visibility_enum import CreateGroupVisibilityEnum
from restapisdk.models.tspublic_rest_v_2_group_adduser_request \
    import TspublicRestV2GroupAdduserRequest
from restapisdk.models.tspublic_rest_v_2_group_create_request \
    import TspublicRestV2GroupCreateRequest
from restapisdk.models.user_name_and_id_input import UserNameAndIDInput

from Python.UserAndGroupCrudSample.UserAndGroupCurd.ControllerBase import ControllerBase


class GroupMethods(ControllerBase):
    def CreateNewGroup(self, client, groupName):

        groupController = client.group
        body = TspublicRestV2GroupCreateRequest()
        body.name = groupName
        body.display_name = 'sample group1'
        body.visibility = CreateGroupVisibilityEnum.DEFAULT
        body.description = 'sample group1'
        body.privileges = \
            [CreateGroupPrivilegesEnum.ADMINISTRATION, CreateGroupPrivilegesEnum.DEVELOPER]

        body.mtype = CreateGroupTypeEnum.LOCAL_GROUP
        count = 0
        result = ''
        try:
            result = groupController.get_group(groupName)
            count = 1
        except:
            print('')
        finally:
            if count == 0:
                result = groupController.create_group(body)

        return result

    def AddSomeUserToGroup(self, client, userCreateBody, groupName):
        groupController = client.group
        groupBody = TspublicRestV2GroupAdduserRequest()
        groupBody.name = groupName
        groupBody.users = []

        groupBody.users.append(UserNameAndIDInput())
        for body in userCreateBody:
            groupBody.users[0].name = body.name
            result = groupController.add_users_to_group(groupBody)
        return result
